### PR TITLE
[SYCL] Use `static` address space cast for `atomic_ref` ctor in SPIR-V path

### PR DESCRIPTION
**Pull request by aelovikov-intel**
From SYCL 2020 specification:

> The sycl::atomic_ref class also has a template parameter AddressSpace,
> which allows the application to make an assertion about the address
> space of the object of type T that it references. The default value
> for this parameter is access::address_space::generic_space, which
> indicates that the object could be in either the global or local
> address spaces. If the application knows the address space, it can set
> this template parameter to either access::address_space::global_space
> or access::address_space::local_space as an assertion to the
> implementation. Specifying the address space via this template
> parameter may allow the implementation to perform certain
> optimizations. Specifying an address space that does not match the
> object’s actual address space results in undefined behavior

We use `ext::oneapi::experimental::static_address_cast` to do that. It's
not implemented for CUDA/HIP yet, that path continues using
`sycl::address_space_cast` that performs runtime checks:

> An implementation must return nullptr if the run-time value of pointer
> is not compatible with Space, and must issue a compiletime diagnostic
> if the deduced address space for pointer is not compatible with Space.

### CODE REVIEW

File: sycl/test/check_device_code/atomic_ref.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+// RUN: %clangxx -O3 -fsycl -fsycl-device-only -fno-discard-value-names -S -emit-llvm -fno-sycl-instrument-device-code  -o - %s | FileCheck %s
+
+#include <sycl/sycl.hpp>
+
+// CHECK-LABEL: define dso_local spir_func noundef i32 @_Z17atomic_ref_globalRi(
+// CHECK-SAME: ptr addrspace(4) noundef align 4 dereferenceable(4) [[I:%.*]]) local_unnamed_addr #[[ATTR0:[0-9]+]] !srcloc [[META6:![0-9]+]] !sycl_fixed_targets [[META7:![0-9]+]] {
+// CHECK-NEXT:  [[ENTRY:.*:]]
+// CHECK-NEXT:    [[CALL_I_I_I_I_I_I:%.*]] = tail call spir_func noundef ptr addrspace(1) @_Z33__spirv_GenericCastToPtr_ToGlobalPvi(ptr addrspace(4) noundef [[I]], i32 noundef 5) #[[ATTR3:[0-9]+]]
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/atomic_ref.cpp:9]: One can review individual commits in this PR to see before/after change.

**Comment by maarquitos14** [comment]
@aelovikov-intel I see you marked this as draft. Is this good to review or is it still work in progress?

**Comment by aelovikov-intel** [comment]
> @aelovikov-intel I see you marked this as draft. Is this good to review or is it still work in progress?

Ready now. I had to wait until https://github.com/intel/llvm/pull/15394 is merged and this is updated to include it via `origin/sycl`.

File: sycl/include/sycl/atomic_ref.hpp [hunk 2]
```diff
@@ -157,8 +158,16 @@ class atomic_ref_base {
   }
 
 #ifdef __SYCL_DEVICE_ONLY__
+#if defined(__SPIR__)
```
Comment (rolandschulz) [inline sycl/include/sycl/atomic_ref.hpp:161]: Why is feature test macro `SYCL_EXT_ONEAPI_ADDRESS_CAST` not defined for the `__SPIR__` case?

**Review by rolandschulz — APPROVED** [review]

File: sycl/include/sycl/atomic_ref.hpp (hunk 2 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/atomic_ref.hpp:161]: Extension implementation seems to have missed that completely. I'm making other changes to the extension, will work on that outside this PR.

**Pull request merged**